File: src/mapping_node.cpp
#include "phi_p3dx_mapping/mapping_node.hpp"
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>

MappingNode::MappingNode(const std::string &node_name, double timer_period)
  : Node(node_name),
    x_(0.0), y_(0.0), theta_(0.0),
    resolution_(0.05), origin_x_cells_(0), origin_y_cells_(0)
{
  // Publisher: mapa global
  rclcpp::QoS map_qos(rclcpp::KeepLast(1));
  map_qos.transient_local();
  map_pub_ = this->create_publisher<nav_msgs::msg::OccupancyGrid>("map", map_qos);

  // Subscriber: odometria
  odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
    "odom", 10,
    std::bind(&MappingNode::odom_callback, this, std::placeholders::_1));

  // Inicializa mapa 20x20m com resolução de 0.05m
  init_map(20.0, 20.0, 0.05);

  RCLCPP_INFO(this->get_logger(), "[%s] iniciado", node_name.c_str());
}

void MappingNode::init_map(double map_width_m, double map_height_m, double resolution)
{
  resolution_ = resolution;
  int width_cells = (int)(map_width_m / resolution_);
  int height_cells = (int)(map_height_m / resolution_);

  map_msg_.header.frame_id = "map";
  map_msg_.info.resolution = resolution_;
  map_msg_.info.width = width_cells;
  map_msg_.info.height = height_cells;
  
  // A origem do mapa fica no canto inferior esquerdo, então para que (0,0) seja o centro:
  map_msg_.info.origin.position.x = -map_width_m / 2.0;
  map_msg_.info.origin.position.y = -map_height_m / 2.0;
  map_msg_.info.origin.position.z = 0.0;
  map_msg_.info.origin.orientation.w = 1.0;

  origin_x_cells_ = width_cells / 2;
  origin_y_cells_ = height_cells / 2;

  // Preenche mapa com células desconhecidas
  map_msg_.data.assign(width_cells * height_cells, -1);
}

void MappingNode::odom_callback(const nav_msgs::msg::Odometry::SharedPtr msg)
{
  x_ = msg->pose.pose.position.x;
  y_ = msg->pose.pose.position.y;

  tf2::Quaternion q(
    msg->pose.pose.orientation.x,
    msg->pose.pose.orientation.y,
    msg->pose.pose.orientation.z,
    msg->pose.pose.orientation.w);
  tf2::Matrix3x3 m(q);
  double roll, pitch;
  m.getRPY(roll, pitch, theta_);

  on_odom();
}

std::tuple<int, int> MappingNode::meters_to_cells(double wx, double wy) const
{
  int mx = (int)((wx - map_msg_.info.origin.position.x) / resolution_);
  int my = (int)((wy - map_msg_.info.origin.position.y) / resolution_);
  return {mx, my};
}

void MappingNode::publish_map()
{
  map_msg_.header.stamp = this->now();
  map_pub_->publish(map_msg_);
}

File: phi_p3dx_mapping/mapping_node.py
import math
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, QoSProfile
from nav_msgs.msg import Odometry, OccupancyGrid

def yaw_from_quaternion(x: float, y: float, z: float, w: float) -> float:
    return math.atan2(2.0 * (w * z + x * y), 1.0 - 2.0 * (y * y + z * z))

class MappingNode(Node):
    """
    Classe base para nós de mapeamento em ROS2.
    Gerencia odometria e publicação de OccupancyGrid.
    """
    def __init__(self, node_name: str = 'mapping_node', timer_period: float = 0.1):
        super().__init__(node_name)

        # Estado do robô
        self.x: float = 0.0
        self.y: float = 0.0
        self.theta: float = 0.0

        # Mapa
        self.map_msg = OccupancyGrid()
        self.resolution = 0.05

        # Publisher do mapa
        map_qos = QoSProfile(
            depth=1,
            durability=DurabilityPolicy.TRANSIENT_LOCAL
        )
        self.map_pub = self.create_publisher(OccupancyGrid, '/map', map_qos)

        # Subscriber de odometria
        self.create_subscription(Odometry, '/odom', self._cb_odom, 10)

        # Timer
        self.create_timer(timer_period, self._control_loop)

        # Inicializa o mapa com 20x20 metros
        self.init_map(20.0, 20.0, 0.05)

        self.get_logger().info(f'[{node_name}] iniciado.')

    def init_map(self, map_width_m: float, map_height_m: float, resolution: float):
        self.resolution = resolution
        width_cells = int(map_width_m / self.resolution)
        height_cells = int(map_height_m / self.resolution)

        self.map_msg.header.frame_id = 'map'
        self.map_msg.info.resolution = self.resolution
        self.map_msg.info.width = width_cells
        self.map_msg.info.height = height_cells
        
        self.map_msg.info.origin.position.x = -map_width_m / 2.0
        self.map_msg.info.origin.position.y = -map_height_m / 2.0
        self.map_msg.info.origin.position.z = 0.0
        self.map_msg.info.origin.orientation.w = 1.0

        self.map_msg.data = [-1] * (width_cells * height_cells)

    def _cb_odom(self, msg: Odometry) -> None:
        self.x = msg.pose.pose.position.x
        self.y = msg.pose.pose.position.y
        q = msg.pose.pose.orientation
        self.theta = yaw_from_quaternion(q.x, q.y, q.z, q.w)
        self.on_odom()

    def world_to_map(self, wx: float, wy: float) -> tuple:
        mx = int((wx - self.map_msg.info.origin.position.x) / self.resolution)
        my = int((wy - self.map_msg.info.origin.position.y) / self.resolution)
        return (mx, my)

    def publish_map(self) -> None:
        self.map_msg.header.stamp = self.get_clock().now().to_msg()
        self.map_pub.publish(self.map_msg)

    def on_odom(self) -> None:
        pass

    def _control_loop(self) -> None:
        pass
